psydac/feec/global_projectors.py

```diff
@@ -0,0 +1,472 @@
+        coeffs : 1D array_like
+            Finite element coefficients obtained by projection.
+        '''
+
+        # build the rhs
+        self.func(*self.args, fun)
+
+        if len(self.N)==1:
+            rhs = self.rhs.toarray()
+            coeffs = array_to_stencil(self.N[0].solve(rhs), self.space.vector_space)
+            coeffs.update_ghost_regions()
+        else:
+            coeffs = kronecker_solve(solvers = self.N, rhs = self.rhs)
+
+        return FemField(self.space, coeffs=coeffs)
+
+class Projector_Hcurl:
+
+    def __init__(self, Hcurl, n_quads=None):
+
+        if n_quads:
+            uw = [gauss_legendre( k-1 ) for k in n_quads]
+            uw = [(u[::-1], w[::-1]) for u,w in uw]
+        else:
+            uw = [(V.quad_grids[i].quad_rule_x,V.quad_grids[i].quad_rule_w) for i,V in enumerate(Hcurl.spaces)]
+
+        dim = len(n_quads)
+
+        self.space  = Hcurl
+        self.rhs    = BlockVector(Hcurl.vector_space)
+
+        for V in Hcurl.spaces:
+            V.init_interpolation()
+            V.init_histopolation()
+
+        if dim == 3:
+
+            # 1D spline spaces (B-splines of degree p and M-splines of degree p-1)
+            Ns = [Hcurl.spaces[1].spaces[0], Hcurl.spaces[0].spaces[1], Hcurl.spaces[0].spaces[2]]
+            Ds = [Hcurl.spaces[0].spaces[0], Hcurl.spaces[1].spaces[1], Hcurl.spaces[2].spaces[2]]
+
+            # Package 1D interpolators and 1D histopolators for 3D Kronecker solver
+            self.DNN = [Ds[0]._histopolator, Ns[1]._interpolator, Ns[2]._interpolator]
+            self.NDN = [Ns[0]._interpolator, Ds[1]._histopolator, Ns[2]._interpolator]
+            self.NND = [Ns[0]._interpolator, Ns[1]._interpolator, Ds[2]._histopolator]
+
+            # Interpolation points
+            intp_x = [V.greville for V in Ns]
+
+            # Quadrature points and weights
+            quads = [quadrature_grid(V.histopolation_grid, u, w) for V,(u,w) in zip(Ds, uw)]
+            quad_x, quad_w = list(zip(*quads))
+
+            # Arrays of degrees of freedom (to be computed) as slices of RHS vector
+            slices = tuple(slice(p, -p) for p in Hcurl.spaces[0].vector_space.pads)
+            dofs   = [x._data[slices] for x in self.rhs]
+
+            # Store data in object
+            self.args = (*intp_x, *quad_x, *quad_w, *dofs)
+            self.func = evaluate_dof_1form_3d
+            self.Ns = Ns
+            self.Ds = Ds
+
+        else:
+            raise NotImplementedError('only 3d is available')
+
+    # ======================================
+    def __call__(self, fun):
+
+        # build the rhs
+        self.func(*self.args, *fun)
+
+        self.rhs.update_ghost_regions()
+
+        coeffs    = BlockVector(self.space.vector_space)
+        coeffs[0] = kronecker_solve(solvers = self.DNN, rhs = self.rhs[0])
+        coeffs[1] = kronecker_solve(solvers = self.NDN, rhs = self.rhs[1])
+        coeffs[2] = kronecker_solve(solvers = self.NND, rhs = self.rhs[2])
```

Comment:
Same comment as above, `out=coeffs[i]` with `i=0,1,2` should be passed as an argument.

Reply:
the coeffs[i] is not allocated so it's not gonna work